File: scripts/analysis/grouping_audit.py
#!/usr/bin/env python3
"""
Grouping Audit Script
Analyzuje reálná data z articles.json a simuluje topic grouping
pro detekci false positives a duplicitních zdrojů
"""

import json
import re
from collections import defaultdict
from datetime import datetime
from pathlib import Path
from typing import List, Dict, Set, Tuple

def normalize_title_for_key(title: str) -> str:
    """Normalizuje title pro výpočet klíče tématu"""
    if not title or not isinstance(title, str):
        return ""
    
    # Lowercase
    normalized = title.lower()
    
    # Odstranění diakritiky
    replacements = {
        'á': 'a', 'à': 'a', 'ä': 'a',
        'é': 'e', 'è': 'e', 'ě': 'e',
        'í': 'i', 'ì': 'i',
        'ó': 'o', 'ò': 'o', 'ö': 'o',
        'ú': 'u', 'ù': 'u', 'ů': 'u', 'ü': 'u',
        'ý': 'y',
        'č': 'c', 'ď': 'd', 'ň': 'n', 'ř': 'r',
        'š': 's', 'ť': 't', 'ž': 'z',
    }
    for old, new in replacements.items():
        normalized = normalized.replace(old, new)
    
    # Odstranění interpunkce a speciálních znaků
    normalized = re.sub(r'[^\w\s]', ' ', normalized)
    # Odstranění čísel (samostatné)
    normalized = re.sub(r'\b\d+\b', ' ', normalized)
    # Redukce whitespace
    normalized = re.sub(r'\s+', ' ', normalized).strip()
    
    # Odstranění "měkkých" stop slov
    soft_stop_words = {"video", "zive", "aktualne", "live", "breaking"}
    words = normalized.split()
    filtered = [w for w in words if len(w) > 2 and w not in soft_stop_words]
    
    return " ".join(filtered)

def compute_topic_key(article: Dict) -> str:
    """Vypočítá klíč tématu z článku"""
    if not article:
        return None
    
    title = article.get("title") or article.get("headline") or article.get("name") or ""
    normalized_title = normalize_title_for_key(title)
    
    if len(normalized_title) < 10:
        topic = (article.get("topic") or "").lower().strip()
        section = (article.get("section") or "").lower().strip()
        if topic or section:
            return f"{topic}||{section}"
    
    return normalized_title or None

def merge_sources_dedup(sources_list: List[List[Dict]]) -> List[Dict]:
    """Sloučí a deduplikuje sources z více článků"""
    seen = set()
    merged = []
    
    for sources in sources_list:
        if not isinstance(sources, list):
            continue
        
        for source in sources:
            if not source or not isinstance(source, dict):
                continue
            
            name = str(source.get("name") or source.get("title") or "").strip()
            url = str(source.get("url") or source.get("link") or "").strip()
            
            if not name or not url:
                continue
            
            key = f"{url.lower()}||{name.lower()}"
            if key in seen:
                continue
            
            seen.add(key)
            merged.append({"name": name, "url": url})
    
    return merged

def group_articles_by_topic(articles: List[Dict], hours: int) -> List[Dict]:
    """Seskupí články podle tématu s časovým oknem"""
    if not articles or hours <= 0:
        return articles
    
    groups = {}
    
    # Seřadit podle publishedAt (ASC)
    sorted_articles = sorted(articles, key=lambda a: 
        datetime.fromisoformat(a.get("publishedAt", a.get("date", a.get("published", "1970-01-01T00:00:00Z"))).replace("Z", "+00:00")).timestamp())
    
    for article in sorted_articles:
        topic_key = compute_topic_key(article)
        if not topic_key:
            continue
        
        if topic_key not in groups:
            published_at = article.get("publishedAt") or article.get("date") or article.get("published") or ""
            try:
                first_time = datetime.fromisoformat(published_at.replace("Z", "+00:00")).timestamp()
            except:
                first_time = 0
            
            groups[topic_key] = {
                "primary": article,
                "related": [],
                "first_time": first_time,
                "time_window_end": first_time + (hours * 3600),
            }
        else:
            group = groups[topic_key]
            published_at = article.get("publishedAt") or article.get("date") or article.get("published") or ""
            try:
                article_time = datetime.fromisoformat(published_at.replace("Z", "+00:00")).timestamp()
            except:
                article_time = 0
            
            if article_time <= group["time_window_end"]:
                group["related"].append(article)
    
    result = []
    
    for topic_key, group in groups.items():
        primary = group["primary"]
        
        all_sources = [
            primary.get("sources", []) if isinstance(primary.get("sources"), list) else [],
        ]
        for related in group["related"]:
            if isinstance(related.get("sources"), list):
                all_sources.append(related["sources"])
        
        merged_sources = merge_sources_dedup(all_sources)
        
        if not primary.get("title") or not primary.get("url") or not isinstance(merged_sources, list):
            result.append(primary)
            continue
        
        grouped_article = {
            **primary,
            "sources": merged_sources,
            "_groupMeta": {
                "relatedCount": len(group["related"]),
                "timeWindow": f"{hours}h",
                "topicKey": topic_key,
            },
        }
        
        result.append(grouped_article)
    
    # Přidat články bez topicKey
    for article in sorted_articles:
        topic_key = compute_topic_key(article)
        if not topic_key:
            result.append(article)
    
    return result

def jaccard_similarity(str1: str, str2: str) -> float:
    """Vypočítá Jaccard podobnost mezi dvěma stringy"""
    tokens1 = set([t for t in str1.lower().split() if len(t) > 2])
    tokens2 = set([t for t in str2.lower().split() if len(t) > 2])
    
    intersection = tokens1 & tokens2
    union = tokens1 | tokens2
    
    return len(intersection) / len(union) if union else 0.0

def analyze_grouping(articles: List[Dict]) -> Dict:
    """Analyzuje seskupování článků"""
    input_count = len(articles)
    grouped = group_articles_by_topic(articles, 12)
    grouped_count = len(grouped)
    
    # Najít skupiny
    group_map = {}
    
    for article in grouped:
        if article.get("_groupMeta") and article["_groupMeta"].get("relatedCount", 0) > 0:
            key = article["_groupMeta"]["topicKey"]
            if key not in group_map:
                group_map[key] = {
                    "key": key,
                    "count": 1 + article["_groupMeta"]["relatedCount"],
                    "primary": article,
                    "related": [],
                    "times": [],
                    "sources": set(),
                    "topics": set(),
                    "sections": set(),
                    "titles": [],
                }
            
            group = group_map[key]
            try:
                time_val = datetime.fromisoformat(article.get("publishedAt", "").replace("Z", "+00:00")).timestamp()
                group["times"].append(time_val)
            except:
                pass
            
            if article.get("topic"):
                group["topics"].add(article["topic"])
            if article.get("section"):
                group["sections"].add(article["section"])
            if article.get("sources"):
                for s in article["sources"]:
                    if s.get("name"):
                        group["sources"].add(s["name"])
            group["titles"].append(article.get("title", ""))
    
    top_groups = sorted(group_map.values(), key=lambda g: g["count"], reverse=True)[:50]
    
    # Detekce podezřelých skupin
    suspicious = []
    
    for group in top_groups:
        issues = []
        
        # Kontrola tokenové podobnosti
        if len(group["titles"]) > 1:
            similarities = []
            for i in range(len(group["titles"])):
                for j in range(i + 1, len(group["titles"])):
                    sim = jaccard_similarity(group["titles"][i], group["titles"][j])
                    similarities.append(sim)
            if similarities:
                avg_sim = sum(similarities) / len(similarities)
                if avg_sim < 0.55:
                    issues.append(f"low_title_similarity:{avg_sim:.2f}")
        
        # Kontrola mixu topic/section
        if len(group["topics"]) > 1:
            issues.append(f"mixed_topics:{','.join(group['topics'])}")
        if len(group["sections"]) > 1:
            issues.append(f"mixed_sections:{','.join(group['sections'])}")
        
        # Kontrola duplicitních zdrojů
        source_names = set()
        source_urls = set()
        dup_names = 0
        dup_urls = 0
        
        if group["primary"].get("sources"):
            for s in group["primary"]["sources"]:
                name_lower = s.get("name", "").lower()
                url_lower = s.get("url", "").lower()
                if name_lower in source_names:
                    dup_names += 1
                if url_lower in source_urls:
                    dup_urls += 1
                if name_lower:
                    source_names.add(name_lower)
                if url_lower:
                    source_urls.add(url_lower)
        
        if dup_names > 0 or dup_urls > 0:
            issues.append(f"duplicate_sources:names={dup_names},urls={dup_urls}")
        
        if issues:
            suspicious.append({
                "key": group["key"],
                "count": group["count"],
                "issues": issues,
                "titles": group["titles"][:5],
                "sources": list(group["sources"])[:10],
            })
    
    return {
        "inputCount": input_count,
        "groupedCount": grouped_count,
        "reduction": input_count - grouped_count,
        "reductionPercent": round((input_count - grouped_count) / input_count * 100, 1) if input_count > 0 else 0,
        "topGroups": [
            {
                "key": g["key"],
                "count": g["count"],
                "timeRange": {
                    "min": datetime.fromtimestamp(min(g["times"])).isoformat() if g["times"] else "",
                    "max": datetime.fromtimestamp(max(g["times"])).isoformat() if g["times"] else "",
                },
                "sources": list(g["sources"])[:10],
                "topics": list(g["topics"]),
                "sections": list(g["sections"]),
                "titles": g["titles"][:5],
            }
            for g in top_groups
        ],
        "suspiciousCount": len(suspicious),
        "suspicious": suspicious[:20],
    }

def main():
    articles_path = Path(__file__).parent.parent.parent / "projects" / "data" / "articles.json"
    
    if not articles_path.exists():
        print(f"❌ File not found: {articles_path}")
        return
    
    with open(articles_path, "r", encoding="utf-8") as f:
        data = json.load(f)
    
    articles = data.get("articles", [])
    if not articles:
        print("❌ No articles found in JSON")
        return
    
    print("🔍 Grouping Audit Report\n")
    print("=" * 80)
    
    report = analyze_grouping(articles)
    
    print(f"\n📊 SUMMARY")
    print(f"Input articles: {report['inputCount']}")
    print(f"Grouped articles: {report['groupedCount']}")
    print(f"Reduction: {report['reduction']} ({report['reductionPercent']}%)")
    print(f"Suspicious groups: {report['suspiciousCount']}")
    
    print(f"\n📈 TOP 10 GROUPS")
    for i, g in enumerate(report["topGroups"][:10], 1):
        key_display = g["key"][:60] + ("..." if len(g["key"]) > 60 else "")
        print(f"\n{i}. Key: \"{key_display}\"")
        print(f"   Count: {g['count']}")
        print(f"   Time range: {g['timeRange']['min']} → {g['timeRange']['max']}")
        sources_display = ", ".join(g["sources"][:5])
        if len(g["sources"]) > 5:
            sources_display += "..."
        print(f"   Sources ({len(g['sources'])}): {sources_display}")
        print(f"   Topics: {', '.join(g['topics']) or 'none'}")
        titles_display = " | ".join([f'"{t[:50]}{"..." if len(t) > 50 else ""}"' for t in g["titles"][:2]])
        print(f"   Titles: {titles_display}")
    
    if report["suspicious"]:
        print(f"\n⚠️  SUSPICIOUS GROUPS ({len(report['suspicious'])})")
        for i, s in enumerate(report["suspicious"][:10], 1):
            key_display = s["key"][:60] + ("..." if len(s["key"]) > 60 else "")
            print(f"\n{i}. Key: \"{key_display}\"")
            print(f"   Count: {s['count']}")
            print(f"   Issues: {', '.join(s['issues'])}")
            titles_display = " | ".join([f'"{t[:40]}{"..." if len(t) > 40 else ""}"' for t in s["titles"][:2]])
            print(f"   Titles: {titles_display}")
    
    print("\n" + "=" * 80)

if __name__ == "__main__":
    main()

File: scripts/analysis/grouping_audit.js
#!/usr/bin/env node
/**
 * Grouping Audit Script
 * Analyzuje reálná data z articles.json a simuluje topic grouping
 * pro detekci false positives a duplicitních zdrojů
 */

const fs = require('fs');
const path = require('path');

// Duplikace logiky z app.js
function normalizeTitleForKey(title) {
  if (!title || typeof title !== "string") return "";
  
  let normalized = title
    .toLowerCase()
    .replace(/[áàä]/g, "a")
    .replace(/[éèě]/g, "e")
    .replace(/[íì]/g, "i")
    .replace(/[óòö]/g, "o")
    .replace(/[úùůü]/g, "u")
    .replace(/[ý]/g, "y")
    .replace(/[č]/g, "c")
    .replace(/[ď]/g, "d")
    .replace(/[ň]/g, "n")
    .replace(/[ř]/g, "r")
    .replace(/[š]/g, "s")
    .replace(/[ť]/g, "t")
    .replace(/[ž]/g, "z")
    .replace(/[^\w\s]/g, " ")
    .replace(/\b\d+\b/g, " ")
    .replace(/\s+/g, " ")
    .trim();
  
  const softStopWords = ["video", "zive", "aktualne", "live", "breaking"];
  const words = normalized.split(/\s+/);
  const filtered = words.filter(w => w.length > 2 && !softStopWords.includes(w));
  
  return filtered.join(" ");
}

function computeTopicKey(article) {
  if (!article) return null;
  
  const title = article.title || article.headline || article.name || "";
  const normalizedTitle = normalizeTitleForKey(title);
  
  if (normalizedTitle.length < 10) {
    const topic = (article.topic || "").toLowerCase().trim();
    const section = (article.section || "").toLowerCase().trim();
    if (topic || section) {
      return `${topic}||${section}`;
    }
  }
  
  return normalizedTitle || null;
}

function mergeSourcesDedup(sourcesArrayList) {
  const seen = new Set();
  const merged = [];
  
  for (const sources of sourcesArrayList) {
    if (!Array.isArray(sources)) continue;
    
    for (const source of sources) {
      if (!source || typeof source !== "object") continue;
      
      const name = String(source.name || source.title || "").trim();
      const url = String(source.url || source.link || "").trim();
      
      if (!name || !url) continue;
      
      const key = `${url.toLowerCase()}||${name.toLowerCase()}`;
      
      if (seen.has(key)) continue;
      seen.add(key);
      
      merged.push({ name, url });
    }
  }
  
  return merged;
}

function groupArticlesByTopic(articles, hours) {
  if (!Array.isArray(articles) || articles.length === 0) return articles;
  if (!Number.isFinite(hours) || hours <= 0) return articles;
  
  const groups = new Map();
  
  const sorted = [...articles].sort((a, b) => {
    const ta = new Date(a.publishedAt || a.date || a.published || 0).getTime();
    const tb = new Date(b.publishedAt || b.date || b.published || 0).getTime();
    return ta - tb;
  });
  
  for (const article of sorted) {
    const topicKey = computeTopicKey(article);
    if (!topicKey) continue;
    
    if (!groups.has(topicKey)) {
      const publishedAt = article.publishedAt || article.date || article.published || "";
      const firstTime = new Date(publishedAt).getTime();
      
      groups.set(topicKey, {
        primary: article,
        related: [],
        firstTime: firstTime,
        timeWindowEnd: firstTime + (hours * 60 * 60 * 1000),
      });
    } else {
      const group = groups.get(topicKey);
      const articleTime = new Date(article.publishedAt || article.date || article.published || 0).getTime();
      
      if (articleTime <= group.timeWindowEnd) {
        group.related.push(article);
      }
    }
  }
  
  const result = [];
  
  for (const [topicKey, group] of groups.entries()) {
    const primary = group.primary;
    
    const allSources = [
      Array.isArray(primary.sources) ? primary.sources : [],
      ...group.related.map(a => Array.isArray(a.sources) ? a.sources : [])
    ];
    const mergedSources = mergeSourcesDedup(allSources);
    
    if (!primary.title || !primary.url || !Array.isArray(mergedSources)) {
      result.push(primary);
      continue;
    }
    
    const groupedArticle = {
      ...primary,
      sources: mergedSources,
      _groupMeta: {
        relatedCount: group.related.length,
        timeWindow: `${hours}h`,
        topicKey: topicKey,
      },
    };
    
    result.push(groupedArticle);
  }
  
  for (const article of sorted) {
    const topicKey = computeTopicKey(article);
    if (!topicKey) {
      result.push(article);
    }
  }
  
  return result;
}

// Jaccard similarity pro detekci false positives
function jaccardSimilarity(str1, str2) {
  const tokens1 = new Set(str1.toLowerCase().split(/\s+/).filter(t => t.length > 2));
  const tokens2 = new Set(str2.toLowerCase().split(/\s+/).filter(t => t.length > 2));
  
  const intersection = new Set([...tokens1].filter(x => tokens2.has(x)));
  const union = new Set([...tokens1, ...tokens2]);
  
  return union.size === 0 ? 0 : intersection.size / union.size;
}

// Hlavní analýza
function analyzeGrouping(articles) {
  const inputCount = articles.length;
  const grouped = groupArticlesByTopic(articles, 12);
  const groupedCount = grouped.length;
  
  // Najít skupiny (články s _groupMeta)
  const groups = grouped.filter(a => a._groupMeta && a._groupMeta.relatedCount > 0);
  const groupMap = new Map();
  
  for (const article of grouped) {
    if (article._groupMeta && article._groupMeta.relatedCount > 0) {
      const key = article._groupMeta.topicKey;
      if (!groupMap.has(key)) {
        groupMap.set(key, {
          key,
          count: 1 + article._groupMeta.relatedCount,
          primary: article,
          related: [],
          times: [],
          sources: new Set(),
          topics: new Set(),
          sections: new Set(),
          titles: [],
        });
      }
      
      const group = groupMap.get(key);
      group.times.push(new Date(article.publishedAt).getTime());
      if (article.topic) group.topics.add(article.topic);
      if (article.section) group.sections.add(article.section);
      if (article.sources) {
        article.sources.forEach(s => group.sources.add(s.name));
      }
      group.titles.push(article.title);
    }
  }
  
  // TOP skupiny
  const topGroups = Array.from(groupMap.values())
    .sort((a, b) => b.count - a.count)
    .slice(0, 50);
  
  // Detekce podezřelých skupin
  const suspicious = [];
  
  for (const group of topGroups) {
    const issues = [];
    
    // Kontrola tokenové podobnosti titulků
    if (group.titles.length > 1) {
      const similarities = [];
      for (let i = 0; i < group.titles.length; i++) {
        for (let j = i + 1; j < group.titles.length; j++) {
          const sim = jaccardSimilarity(group.titles[i], group.titles[j]);
          similarities.push(sim);
        }
      }
      const avgSim = similarities.reduce((a, b) => a + b, 0) / similarities.length;
      if (avgSim < 0.55) {
        issues.push(`low_title_similarity:${avgSim.toFixed(2)}`);
      }
    }
    
    // Kontrola mixu topic/section
    if (group.topics.size > 1) {
      issues.push(`mixed_topics:${Array.from(group.topics).join(',')}`);
    }
    if (group.sections.size > 1) {
      issues.push(`mixed_sections:${Array.from(group.sections).join(',')}`);
    }
    
    // Kontrola duplicitních zdrojů v merged sources
    const sourceNames = new Set();
    const sourceUrls = new Set();
    let dupNames = 0;
    let dupUrls = 0;
    
    if (group.primary.sources) {
      for (const s of group.primary.sources) {
        if (sourceNames.has(s.name?.toLowerCase())) dupNames++;
        if (sourceUrls.has(s.url?.toLowerCase())) dupUrls++;
        if (s.name) sourceNames.add(s.name.toLowerCase());
        if (s.url) sourceUrls.add(s.url.toLowerCase());
      }
    }
    
    if (dupNames > 0 || dupUrls > 0) {
      issues.push(`duplicate_sources:names=${dupNames},urls=${dupUrls}`);
    }
    
    if (issues.length > 0) {
      suspicious.push({
        key: group.key,
        count: group.count,
        issues,
        titles: group.titles.slice(0, 5),
        sources: Array.from(group.sources),
      });
    }
  }
  
  return {
    inputCount,
    groupedCount,
    reduction: inputCount - groupedCount,
    reductionPercent: ((inputCount - groupedCount) / inputCount * 100).toFixed(1),
    topGroups: topGroups.map(g => ({
      key: g.key,
      count: g.count,
      timeRange: {
        min: new Date(Math.min(...g.times)).toISOString(),
        max: new Date(Math.max(...g.times)).toISOString(),
      },
      sources: Array.from(g.sources),
      topics: Array.from(g.topics),
      sections: Array.from(g.sections),
      titles: g.titles.slice(0, 5),
    })),
    suspiciousCount: suspicious.length,
    suspicious,
  };
}

// Main
const articlesPath = path.join(__dirname, '../../projects/data/articles.json');

if (!fs.existsSync(articlesPath)) {
  console.error(`❌ File not found: ${articlesPath}`);
  process.exit(1);
}

const data = JSON.parse(fs.readFileSync(articlesPath, 'utf8'));
const articles = Array.isArray(data.articles) ? data.articles : [];

if (articles.length === 0) {
  console.error('❌ No articles found in JSON');
  process.exit(1);
}

console.log('🔍 Grouping Audit Report\n');
console.log('='.repeat(80));

const report = analyzeGrouping(articles);

console.log(`\n📊 SUMMARY`);
console.log(`Input articles: ${report.inputCount}`);
console.log(`Grouped articles: ${report.groupedCount}`);
console.log(`Reduction: ${report.reduction} (${report.reductionPercent}%)`);
console.log(`Suspicious groups: ${report.suspiciousCount}`);

console.log(`\n📈 TOP 10 GROUPS`);
report.topGroups.slice(0, 10).forEach((g, i) => {
  console.log(`\n${i + 1}. Key: "${g.key.substring(0, 60)}${g.key.length > 60 ? '...' : ''}"`);
  console.log(`   Count: ${g.count}`);
  console.log(`   Time range: ${g.timeRange.min} → ${g.timeRange.max}`);
  console.log(`   Sources (${g.sources.length}): ${g.sources.slice(0, 5).join(', ')}${g.sources.length > 5 ? '...' : ''}`);
  console.log(`   Topics: ${g.topics.join(', ') || 'none'}`);
  console.log(`   Titles: ${g.titles.slice(0, 2).map(t => `"${t.substring(0, 50)}${t.length > 50 ? '...' : ''}"`).join(' | ')}`);
});

if (report.suspicious.length > 0) {
  console.log(`\n⚠️  SUSPICIOUS GROUPS (${report.suspicious.length})`);
  report.suspicious.slice(0, 10).forEach((s, i) => {
    console.log(`\n${i + 1}. Key: "${s.key.substring(0, 60)}${s.key.length > 60 ? '...' : ''}"`);
    console.log(`   Count: ${s.count}`);
    console.log(`   Issues: ${s.issues.join(', ')}`);
    console.log(`   Titles: ${s.titles.slice(0, 2).map(t => `"${t.substring(0, 40)}${t.length > 40 ? '...' : ''}"`).join(' | ')}`);
  });
}

console.log('\n' + '='.repeat(80));
